components/raftstore/src/store/fsm/peer.rs

```diff
@@ -987,36 +988,104 @@ where
         syncer.report_for_self(self_report);
     }
 
-    // Call msg PrepareFlashback to stop the scheduling and RW tasks.
-    // Once called, it will wait for the channel's notification in FlashbackState to
-    // finish. We place a flag in the request, which is checked when the
-    // pre_propose_raft_command is called. Stopping tasks is done by applying
-    // the flashback-only command in this way, But for RW local reads which need
-    // to be considered, we let the leader lease to None to ensure that local reads
-    // are not executed.
+    // Call msg PrepareFlashback is to check if the persistent state is present in
+    // RegionLocalState. Once called, the persistent state and memory state will
+    // first be checked and if neither is found, then a Raft Admin Command will be
+    // proposed to persist the state in the RegionLocalState to lock the region.
+    // Once this Command is submitted, a callback will be invoked to call msg
+    // PrepareFlashback again to wait for the new change to be applied completely.
+    // So we can perform the flashback safely later. We make FlashbackToVersion
+    // become a two-phase request, and lock the region as the first phase.
     fn on_prepare_flashback(&mut self, ch: Sender<bool>) {
+        let region_id = self.region().get_id();
         info!(
             "prepare flashback";
-            "region_id" => self.region().get_id(),
+            "region_id" => region_id,
             "peer_id" => self.fsm.peer.peer_id(),
         );
-        if self.fsm.peer.flashback_state.is_some() {
-            ch.send(false).unwrap();
-            return;
+        // check persist state
+        let state_key = keys::region_state_key(region_id);
+        if let Some(state) = self
+            .ctx
+            .engines
+            .kv
+            .get_msg_cf::<RegionLocalState>(CF_RAFT, &state_key)
```

Comment:
Why check persist state here? Persistent state is used only on start to recover the memory state.

Reply:
I think it's because we can not know whether the current peer is in a flashback state in `PeerFSM` before we read once from the persistent state?

Reply:
you can just check the memory state. Memory state is filled on start-up.

Reply:
I think since the `notify` method will be called after `apply` to send the result of apply to peer fsm.The memory state will be set in the `on_flashback_memory_set` function. So here we need to set the meta information of cf and current region.